### PR TITLE
[SPARK-9725] [SQL] fix serialization of UTF8String across different JVM

### PR DESCRIPTION
**Pull request by davies**
The BYTE_ARRAY_OFFSET could be different in JVM with different configurations (for example, different heap size, 24 if heap > 32G, otherwise 16), so offset of UTF8String is not portable, we should handler that during serialization.

### CODE REVIEW

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -41,9 +40,9 @@
 public final class UTF8String implements Comparable<UTF8String>, Serializable {
 
   @Nonnull
-  private final Object base;
-  private final long offset;
-  private final int numBytes;
+  private Object base;
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Is there a way to still having them as `final`?

Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: See if you can use Externalizable instead.

Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Externalizable still need them to be non-final. Should we use UNSAFE here?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -978,4 +983,22 @@ public UTF8String soundex() {
...
     return UTF8String.fromBytes(sx);
   }
+
+  public void writeExternal(ObjectOutput out) throws IOException {
+    byte[] bytes = getBytes();
+    out.writeInt(bytes.length);
+    out.write(bytes);
+  }
+
+  static {
```
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: what is this for?

**Pull request closed without merging**